From 350d1ea9ca79b923d101670f09cb7f3cced2e69d Mon Sep 17 00:00:00 2001
From: Nich <nctrenco@gmail.com>
Date: Sun, 10 Jun 2018 00:45:51 +0800
Subject: [PATCH 7/7] linker: Make shim reference path absolute

This way, we can filter out non-existent binaries, and ensure we get
its absolute path before matching with get_realpath(). This for one
allows the use of symlinks in TARGET_LD_SHIM_LIBS.

Change-Id: I823815271b3257965534b6b87d8ea36ffb68bc08
---
 linker/linker.cpp | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/linker/linker.cpp b/linker/linker.cpp
index 2b2ba156d..a0dae8a68 100644
--- a/linker/linker.cpp
+++ b/linker/linker.cpp
@@ -733,9 +733,11 @@ void parse_LD_SHIM_LIBS(const char* path) {
   if (path != nullptr) {
     // We have historically supported ':' as well as ' ' in LD_SHIM_LIBS.
     for (const auto& pair : android::base::Split(path, " :")) {
-      size_t pos = pair.find('|');
-      if (pos > 0 && pos < pair.length() - 1) {
-        auto desc = std::pair<std::string, std::string>(pair.substr(0, pos), pair.substr(pos + 1));
+      std::vector<std::string> pieces = android::base::Split(pair, "|");
+      if (pieces.size() != 2) continue;
+      char resolved_path[PATH_MAX];
+      if (realpath(pieces[0].c_str(), resolved_path) != nullptr) {
+        auto desc = std::pair<std::string, std::string>(std::string(resolved_path), pieces[1]);
         g_ld_all_shim_libs.push_back(desc);
       }
     }
-- 
2.17.1

